test/cli/test_main.py

```diff
@@ -237,3 +298,100 @@ def test_cannot_recurse(self):
 
             results = yaml.load(yaml_config)
             assert results == {'FOO': "${VAR1}", 'BAR': [1, 2, 3]}
+
+    @pytest.mark.parametrize(('yaml_config', 'env_vars', 'expected_config'), [
+        # recursive env with root value
+        (
+            """
+            FOO: ${FOO:val_${INDICE:1}}
+            """,
+            {"FOO": 'val_a'},
+            {"FOO": 'val_a'},
+        ),
+        # recursive env with root default and sub value
+        (
+            """
+            FOO: ${FOO:val_${INDICE:1}}
+            """,
+            {"INDICE": 'b'},
+            {"FOO": 'val_b'},
+        ),
+        # recursive env requiring data
+        (
+            """
+            FOO: ${FOO:val_${INDICE}}
+            """,
+            {"INDICE": 'b'},
+            {"FOO": 'val_b'},
+        ),
+
+        # default data with bracket
+        (
+            """
+            FOO: ${FOO:"{name} {age}"}
+            """,
+            {},
+            {"FOO": '{name} {age}'},
+        ),
+        # default data with bracket
+        (
+            """
+            FOO: ${FOO:"{name} {age}"}
+            """,
+            {"FOO": '"{surname}"'},
+            {"FOO": '{surname}'},
+        ),
+    ])
+    @pytest.mark.skipif(not has_regex_module,
+                        reason='0 support for nested env without regex module')
+    def test_environment_vars_recursive_in_config(
+            self, yaml_config, env_vars, expected_config
+    ):  # pragma: no cover
+        setup_yaml_parser()
+
+        with patch.dict('os.environ'):
+            for key, val in env_vars.items():
+                os.environ[key] = val
+
+            results = yaml.load(yaml_config)
+            assert results == expected_config
+
+    @pytest.mark.parametrize(('yaml_config', 'env_vars', 'expected_config'), [
+        # recursive env with root value
+        (
+            """
+            FOO: ${FOO:val_${INDICE:1}}
+            """,
+            {"FOO": 'val_a'},
+            {"FOO": 'val_a}'},
+        ),
+        # recursive env with root default and sub value
+        (
+            """
+            FOO: ${FOO:val_${INDICE:1}}
+            """,
+            {"INDICE": 'b'},
+            {"FOO": 'val_${INDICE:1}'},
+        ),
+        # recursive env requiring data
+        (
+            """
+            FOO: ${FOO:${INDICE}_val}
+            """,
+            {"INDICE": 'b'},
+            {"FOO": '${INDICE_val}'},
```

Comment:
Shouldn't the expected value here be `${INDICE}_val`? Note the extra `}`

Reply:
the ENV_VAR_MATCHER regex behavior is to match until the first ``}``. this split the given config into: 
``${FOO:${INDICE}``**+**``_val}``
with ``FOO`` having default value of ``${INDICE`` concatenated with raw string ``_val}``

Reply:
Is this desirable behaviour though?

Actually I would expect this case to evaluate as

```
FOO: b_val
```

Reply:
Comparing this case to the one above in `test_environment_vars_recursive_in_config`:

```
# test_environment_vars_recursive_in_config:
# recursive env requiring data
(
    """
    FOO: ${FOO:val_${INDICE}}
    """,
    {"INDICE": 'b'},
    {"FOO": 'val_b'},  # << expected value "val_b"
),
```

```
# test_unhandled_recursion:
# recursive env requiring data
(
     """
     FOO: ${FOO:${INDICE}_val}
     """,
     {"INDICE": 'b'},
     {"FOO": '${INDICE_val}'},   # << expected value should be "b_val"
)
```

What's the difference?

Reply:
I have added test for both old and new behaviors. the behaviors described is for the old behavior ``@pytest.mark.skipif(has_regex_module,`` => don't run if we are in the env with regex, so this run only without regex, and the old behavior is expected.

=> the difference is in the ``@pytest.mark.skipif(==>no<==t has_regex_module,``

Reply:
Ahh, I understand now. Sorry for the confusion